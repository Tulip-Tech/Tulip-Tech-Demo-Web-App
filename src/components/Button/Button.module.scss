@use 'src/styles/variables';
@use 'src/styles/theme';
@use 'src/styles/mixins/responsive';

$small-button-height: 28px;
$medium-button-height: 36px;
$large-button-height: 40px;

.button {
  position: relative;
  display: inline-flex;
  justify-content: inherit;
  align-items: center;
  height: $medium-button-height;
  padding: 0 16px;
  overflow: hidden;

  font-family: var(--body-alt-font-family);
  font-weight: var(--body-font-weight-bold);
  font-size: 16px;
  line-height: 18px;
  text-align: center;
  text-decoration: none;
  border-radius: 4px;
  outline: none;
  cursor: pointer;
  transition: background-color 0.1s ease, transform 0.1s ease;

  @media (hover: hover) and (pointer: fine) {
    &:not(.disabled) {
      &:hover,
      &:focus {
        z-index: 1;
        transform: scale(1.1);
      }

      &:focus:not(:focus-visible):not(:hover) {
        transform: scale(1);
      }

      &:focus-visible {
        transform: scale(1.1);
      }
    }
  }

  &.disabled {
    cursor: default;
    opacity: 0.5;
  }

  &.small {
    height: $small-button-height;
  }

  &.large {
    height: $large-button-height;
  }

  &.default {
    color: currentColor;
    background-color: theme.$btn-default-bg;
  }

  &.primary {
    color: var(--highlight-contrast-color, theme.$btn-primary-color);
    background-color: var(--highlight-color, theme.$btn-primary-bg);
  }

  &.headerbtn {
    width: max-content;
    padding: 5px 5px 5px 8px;

    color: variables.$white;
    font-weight: 800;
    font-size: 14px;
    background-color: var(--highlight-color, theme.$btn-primary-bg);
    border-radius: 1px;
  }
  &.outlined {
    width: auto;
    border: 1px solid rgba(255, 255, 255, 0.3);

    &:not(.disabled) {
      &.active,
      &:focus {
        color: var(--highlight-contrast-color, theme.$btn-primary-color);
        background-color: var(--highlight-color, theme.$btn-primary-bg);
        border-color: var(--highlight-color, theme.$btn-primary-bg);
      }
    }
  }

  &.text {
    background: none;
    opacity: 0.7;

    &:not(.disabled) {
      &.active,
      &:focus {
        opacity: 1;
      }

      &:hover {
        z-index: 1;
        background: theme.$btn-default-bg;
        opacity: 1;
      }
    }
  }

  &.fullWidth {
    justify-content: center;
    width: 100%;

    @media (hover: hover) and (pointer: fine) {
      &:hover:not(.disabled) {
        transform: scale(1.04);
      }
    }
  }

  > svg {
    width: 18px;
    height: 18px;
  }
}

.startIcon {
  display: flex;
  align-items: center;
  height: 100%;
  margin-right: 11px;

  > svg {
    width: 20px;
    height: 20px;
    fill: currentColor;

    @include responsive.tablet-only() {
      width: 18px;
      height: 18px;
    }
  }
}

.hidden {
  visibility: hidden;
}

.centerAbsolute {
  position: absolute;
  right: 0;
  left: 0;
  margin: auto;
  transform: translate(-5px, -5px);
}
