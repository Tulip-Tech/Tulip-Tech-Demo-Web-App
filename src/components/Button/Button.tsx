import React, { MouseEventHandler } from 'react';
import classNames from 'classnames';
import { NavLink } from 'react-router-dom';

import styles from './Button.module.scss';

import Spinner from '#components/Spinner/Spinner';

type Color = 'default' | 'primary' | 'headerbtn';

type Variant = 'contained' | 'outlined' | 'text';

type Props = {
  children?: React.ReactNode;
  label: string;
  active?: boolean;
  color?: Color;
  fullWidth?: boolean;
  startIcon?: JSX.Element;
  variant?: Variant;
  onClick?: MouseEventHandler<HTMLButtonElement>;
  tabIndex?: number;
  size?: 'small' | 'medium' | 'large';
  to?: string;
  role?: string;
  className?: string;
  type?: 'button' | 'submit' | 'reset';
  disabled?: boolean;
  busy?: boolean;
  id?: string;
} & React.AriaAttributes;

const Button: React.FC<Props> = ({
  label,
  children,
  color = 'default',
  startIcon,
  fullWidth = false,
  active = false,
  variant = 'outlined',
  size = 'medium',
  disabled,
  busy,
  type,
  to,
  onClick,
  className,
  ...rest
}: Props) => {
  const buttonClassName = (isActive: boolean) =>
    classNames(styles.button, className, styles[color], styles[variant], {
      [styles.active]: isActive,
      [styles.fullWidth]: fullWidth,
      [styles.large]: size === 'large',
      [styles.small]: size === 'small',
      [styles.disabled]: disabled,
    });

  const content = (
    <>
      {startIcon && <div className={styles.startIcon}>{startIcon}</div>}
      {<span className={classNames(styles.buttonLabel, { [styles.hidden]: busy }) || undefined}>{label}</span>}
      {children}
      {busy && <Spinner className={styles.centerAbsolute} size={'small'} />}
    </>
  );

  if (to) {
    return (
      <NavLink className={({ isActive }) => buttonClassName(isActive)} to={to} {...rest} end>
        {content}
      </NavLink>
    );
  }

  return (
    <button className={buttonClassName(active)} onClick={onClick} type={type} disabled={disabled} aria-disabled={disabled} {...rest}>
      {content}
    </button>
  );
};
export default Button;
