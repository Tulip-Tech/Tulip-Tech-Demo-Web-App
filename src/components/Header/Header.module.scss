@use 'src/styles/variables';
@use 'src/styles/theme';
@use 'src/styles/mixins/responsive';
@use 'src/styles/mixins/utils';

//
// Header
// --------------------------------

.header {
  height: 50px;
  padding: 10px variables.$base-spacing * 4;
  color: var(--header-contrast-color, variables.$white);
  background: var(--header-background, transparent);

  //
  // Make header fixed
  //
  &.fixed {
    position: fixed;
    z-index: variables.$header-z-index;
    width: 100%;
  }

  //
  // Make header static
  //
  &.static {
    position: static;
    width: 100%;
  }
}

//
// Header container
//
.container {
  position: relative;
  display: flex;
  flex-direction: row;
  height: 100%;
}

//
// Header menu
//
.menu {
  display: none;
}

//
// Header menu
//
.iconButton {
  width: 36px;
  height: 36px;

  > svg {
    fill: currentColor;
  }
}

//
// Header brand
//
.brand {
  align-self: center;
}

//
// Header navigation
//
.nav {
  display: inline-block;
  flex: 1;
  align-items: center;

  > a {
    height: 36px;
    min-height: 36px;
    margin: 0 6px;
    font-weight: var(--body-font-weight-bold);
    font-size: 18px;
  }
}

//
// Header search
//
.search {
  max-width: 180px;
}

//
// Header mobile search
//
.mobileSearch {
  position: absolute;
  right: 0;
  left: 0;
  display: flex;

  > div:first-child {
    flex: 1;
  }
}

//
// Header modifier when mobile search is active
//
.mobileSearchActive {
  .menu,
  .brand {
    display: none;
  }
}

.buttonContainer {
  > button:first-child {
    margin-right: calc(#{variables.$base-spacing} / 2);
    margin-left: variables.$base-spacing;
  }
}

.userMenuButton {
  margin-left: 8px;
}

.brandCentered {
  .brand {
    text-align: center;
  }
}
//
// mediaQueries
// --------------------------------

@include responsive.mobile-and-tablet() {
  .header {
    height: variables.$header-height-mobile;
    padding: 10px variables.$base-spacing * 2;
  }

  .menu {
    display: block;
  }

  .brand {
    flex: 1;
    margin: 0;
    margin-left: 8px;
    text-align: left;
  }

  .nav {
    display: none;
  }

  .search {
    text-align: right;
  }
}

@include responsive.mobile-only() {
  .header {
    padding: 10px variables.$base-spacing;
  }

  .menu {
    width: 40px;
  }

  .search {
    width: 40px;
  }
}
