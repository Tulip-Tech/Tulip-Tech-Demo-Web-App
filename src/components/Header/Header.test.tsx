import React from 'react';
import { render } from '@testing-library/react';

import Header from './Header';

import Button from '#components/Button/Button';

vi.mock('react-router-dom', () => ({
  NavLink: () => 'a',
}));

describe('<Header />', () => {
  test('renders header', () => {
    const playlistMenuItems = [<Button key="key" label="Home" to="/" />];
    const { container } = render(
      <Header
        onMenuButtonClick={vi.fn()}
        searchBarProps={{
          query: '',
          onQueryChange: vi.fn(),
        }}
        searchEnabled
        searchActive={false}
        onSearchButtonClick={vi.fn()}
        onCloseSearchButtonClick={vi.fn()}
        onLoginButtonClick={vi.fn()}
        userMenuOpen={false}
        toggleUserMenu={vi.fn()}
        isLoggedIn={false}
        canLogin={true}
        showPaymentsMenuItem={true}
      >
        {playlistMenuItems}
      </Header>,
    );

    expect(container).toMatchSnapshot();
  });
});
