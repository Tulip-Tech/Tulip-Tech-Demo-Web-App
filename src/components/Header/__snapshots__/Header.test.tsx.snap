// Vitest Snapshot v1

exports[`<Header /> > renders header 1`] = `
<div>
  <header
    class="_header_f4f7a7 _static_f4f7a7"
  >
    <div
      class="_container_f4f7a7"
    >
      <div
        class="_menu_f4f7a7"
      >
        <div
          aria-label="open_menu"
          class="_iconButton_0fef65 _iconButton_f4f7a7"
          role="button"
          tabindex="0"
        >
          <svg
            aria-hidden="true"
            class="_icon_1e9999"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </g>
          </svg>
        </div>
      </div>
      <nav
        aria-label="menu"
        class="_nav_f4f7a7"
      >
        a
      </nav>
      <div
        class="_search_f4f7a7"
      >
        <div
          class="_searchBar_3bb1d7"
        >
          <svg
            aria-hidden="true"
            class="_icon_1e9999 _icon_3bb1d7"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </g>
          </svg>
          <input
            aria-label="search_bar.search_label"
            class="_input_3bb1d7"
            placeholder="search_bar.search_placeholder"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="_buttonContainer_f4f7a7"
      >
        <button
          class="_button_f8f296 _default_f8f296 _outlined_f8f296"
        >
          <span>
            sign_in
          </span>
        </button>
        <button
          class="_button_f8f296 _primary_f8f296"
        >
          <span>
            sign_up
          </span>
        </button>
      </div>
    </div>
  </header>
</div>
`;
